girder/utility/acl_mixin.py

```diff
@@ -113,51 +186,58 @@ def requireAccess(self, doc, user=None, level=AccessType.READ):
                 userid = str(user.get('_id', ''))
             else:
                 userid = None
-            raise AccessException("%s access denied for %s %s (user %s)." %
-                                  (perm, self.name, doc.get('_id', 'unknown'),
-                                   userid))
+            raise AccessException('%s access denied for %s %s (user %s).' %
+                                  (perm, self.name, doc.get('_id', 'unknown'), userid))
 
     def requireAccessFlags(self, doc, user=None, flags=None):
         """
         See the documentation of AccessControlledModel.requireAccessFlags, which this wraps.
         """
-        if not flags:
-            return
+        if not self.hasAccessFlags(doc, user, flags):
+            if user:
+                uid = str(user.get('_id', ''))
+            else:
+                uid = None
 
-        resource = self.model(self.resourceColl).load(doc[self.resourceParent], force=True)
-        return self.model(self.resourceColl).requireAccessFlags(resource, user, flags)
+            raise AccessException('Access denied for %s %s (user %s).' %
+                                  (self.name, doc.get('_id', 'unknown'), uid))
 
     def filterResultsByPermission(self, cursor, user, level, limit=0, offset=0,
                                   removeKeys=(), flags=None):
         """
-        Yields filtered results from the cursor based on the access control
-        existing for the resourceParent.
+        Yields filtered results from the cursor based on the access control for the resource.
 
         Takes the same parameters as
         :py:func:`girder.models.model_base.AccessControlledModel.filterResultsByPermission`.
         """
-        # Cache mapping resourceIds -> access granted (bool)
+        # Cache mapping (resourceModelName, resourceId) -> access granted (bool)
         resourceAccessCache = {}
 
         def hasAccess(_result):
-            resourceId = _result[self.resourceParent]
-
-            # return cached check if it exists
-            if resourceId in resourceAccessCache:
-                return resourceAccessCache[resourceId]
+            # Calling "_resourceModelandId" with "returnType=True" returns the type of the resource
+            # as a hashable string or list, and doesn't call ModelImporter (so it's fast)
+            resourceType, resourceId = self._resourceModelandId(_result, returnType=True)
```

Comment:
Isn't `resourceId` guaranteed (by Mongo) to be distinct regardless of the `resourceType`?  As such, isn't `resourceId` sufficient for the cache key?

Reply:
I think so, but since this is supposed to also handle plugin models, I see no reason to make that assumption unnecessarily, given that there's no performance benefit to also including the model type in the cache key.